substrate/frame/identity/src/migration.rs

```diff
@@ -116,11 +156,560 @@ pub mod v1 {
+				MigrationState::FinishedPendingUsernames |
+				MigrationState::CleanupAuthorities(_) => T::WeightInfo::migration_v2_cleanup_authority_step(),
+				MigrationState::FinishedCleanupAuthorities |
+				MigrationState::CleanupUsernames(_) => T::WeightInfo::migration_v2_cleanup_username_step(),
+				MigrationState::Finished => Weight::zero(),
+			}
+		}
+
+		// Migrate one entry from `UsernameAuthorities` to `AuthorityOf`.
+		pub(crate) fn authority_step(maybe_last_key: Option<&T::AccountId>) -> StepResultOf<T> {
+			let mut iter = if let Some(last_key) = maybe_last_key {
+				types_v1::UsernameAuthorities::<T>::iter_from(
+					types_v1::UsernameAuthorities::<T>::hashed_key_for(last_key),
+				)
+			} else {
+				types_v1::UsernameAuthorities::<T>::iter()
+			};
+			if let Some((authority_account, properties)) = iter.next() {
+				let suffix = properties.account_id;
+				let allocation = properties.allocation;
+				let new_properties =
+					AuthorityProperties { account_id: authority_account.clone(), allocation };
+				AuthorityOf::<T>::insert(&suffix, new_properties);
+				MigrationState::Authority(authority_account)
+			} else {
+				MigrationState::FinishedAuthorities
+			}
+		}
+
+		// Migrate one entry from `AccountOfUsername` to `UsernameInfoOf`.
+		pub(crate) fn username_step(maybe_last_key: Option<&Username<T>>) -> StepResultOf<T> {
+			let mut iter = if let Some(last_key) = maybe_last_key {
+				types_v1::AccountOfUsername::<T>::iter_from(
+					types_v1::AccountOfUsername::<T>::hashed_key_for(last_key),
+				)
+			} else {
+				types_v1::AccountOfUsername::<T>::iter()
+			};
+
+			if let Some((username, owner_account)) = iter.next() {
+				let username_info = UsernameInformation {
+					owner: owner_account,
+					provider: Provider::new_with_allocation(),
+				};
+				UsernameInfoOf::<T>::insert(&username, username_info);
+
+				MigrationState::Username(username)
+			} else {
+				MigrationState::FinishedUsernames
+			}
+		}
+
+		// Migrate one entry from `IdentityOf` to `UsernameOf`, if it has a username associated with
+		// it. Remove the entry if there was no real identity associated with the account.
+		pub(crate) fn identity_step(maybe_last_key: Option<HashedKey>) -> StepResultOf<T> {
+			if let Some(last_key) =
+				IdentityOf::<T>::translate_next::<
+					(
+						Registration<
+							BalanceOf<T>,
+							<T as pallet::Config>::MaxRegistrars,
+							<T as pallet::Config>::IdentityInformation,
+						>,
+						Option<Username<T>>,
+					),
+					_,
+				>(maybe_last_key.map(|b| b.to_vec()), |account, (identity, maybe_username)| {
+					if let Some(primary_username) = maybe_username {
+						UsernameOf::<T>::insert(&account, primary_username);
+					}
+					if identity.deposit > BalanceOf::<T>::zero() {
+						Some(identity)
+					} else {
+						None
+					}
+				}) {
+				MigrationState::Identity(last_key.try_into().unwrap())
+			} else {
+				MigrationState::FinishedIdentities
+			}
```

Comment:
I don't understand how this one can work?

in the migration loop it is initially called with `None` after the `FinishedUsernames` state.

But here we just return `FinishedIdentities` when `None` is given.

Reply:
ah I missread, thanks